<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Обзор данных</a></span></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Обработка-пропусков" data-toc-modified-id="Обработка-пропусков-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Обработка пропусков</a></span></li><li><span><a href="#Замена-типа-данных" data-toc-modified-id="Замена-типа-данных-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Замена типа данных</a></span></li><li><span><a href="#Обработка-дубликатов" data-toc-modified-id="Обработка-дубликатов-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Обработка дубликатов</a></span></li><li><span><a href="#Лемматизация" data-toc-modified-id="Лемматизация-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Лемматизация</a></span></li><li><span><a href="#Категоризация-данных" data-toc-modified-id="Категоризация-данных-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Категоризация данных</a></span></li></ul></li><li><span><a href="#Ответьте-на-поставленные-вопросы" data-toc-modified-id="Ответьте-на-поставленные-вопросы-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Ответьте на поставленные вопросы</a></span></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Общий вывод</a></span></li></ul></div>

# Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. 
Цель исследования - дать ответы на каждый из четырех вопросов:
1. Есть ли зависимость между наличием детей и возвратом кредита в срок?
2. Есть ли зависимость между семейным положением и возвратом кредита в срок?
3. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
4. Как разные цели кредита влияют на его возврат в срок?

**Ход исследования**

Входные данные от банка — статистика о платёжеспособности клиентов. О качестве данных ничего не известно. Поэтому перед тем как дать ответы на поставленые вопросы, понадобится обзор данных. Нужно проверить данные на ошибки и оцените их влияние на исследование. Затем, на этапе предобработки нужно исправить самые критичные ошибки данных, а так же категоризировать данные.

Таким образом, исследование пройдёт в три этапа:
 1. Обзор данных.
 2. Предобработка данных.
 3. Ответы на поставленые вопросы

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.


##  Обзор данных


Импортируем библиотеку Pandas

In [1]:
# импортируем библиотеку Pandas
# импортируем библиотеку pymystem3
# импортируем счётчик Counter
import pandas as pd
from pymystem3 import Mystem
from collections import Counter

Прочитаем файл '/datasets/data.csv' и сохраним его в переменной 'bank'

In [2]:
# читаем файл и сохраняем его в bank
bank = pd.read_csv('/datasets/data.csv') 
bank.to_csv('bank.csv', index=False)

Выведем на экран первые 10 строк таблицы

In [3]:
# выводим первые 10 строк таблицы
bank.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

Получим общую информацию о таблице

In [4]:
# получаем информацию о таблице
bank.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Итак в таблице 12 столбцов и данные о 21525 клиенте банка. Типы данных в столбцах: int64, object, float64. 
Согласно документации к данным:
* 'children' — количество детей в семье
* 'days_employed' — общий трудовой стаж в днях
* 'dob_years' — возраст клиента в годах
* 'education' — уровень образования клиента
* 'education_id' — идентификатор уровня образования
* 'family_status' — семейное положение
* 'family_status_id' — идентификатор семейного положения
* 'gender' — пол клиента
* 'income_type' — тип занятости
* 'debt' — имел ли задолженность по возврату кредитов
* 'total_income' — ежемесячный доход
* 'purpose' — цель получения кредита

Количество значений в столбцах различается. Значит, в данных есть пропущенные значения.

**Вывод**

В каждой строке таблицы данные о клиентах банка. Часть колонок описывают социальное положение клиента, часть описывает его финансовое положение и есть отдельная колонка о цели получения кредита. В данных встречаются пропуски. Чтобы ответить на поставление вопросы, нужно устранить проблемы в данных

##  Предобработка данных

###  Обработка пропусков

Посчитаем, сколько в таблице пропущенных значений.

In [5]:
# подсчет пропусков
bank.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

Одинаковое количество пропусков в двух столбцах: 'days_employed' и 'total_income'. Тип данных в обоих столбцах - вещественные числа. Стаж и ежемесячный доход - это количественные переменные.  

Методом isna() найдём все строки с пропусками в столбце 'days_employed'.

In [6]:
# посмотрим на пропуски в столбце 'days_employed' и выведим их на экран.
display(bank[bank['days_employed'].isna()])

children  days_employed  dob_years education  education_id  \
12            0            NaN         65   среднее             1   
26            0            NaN         41   среднее             1   
29            0            NaN         63   среднее             1   
41            0            NaN         50   среднее             1   
55            0            NaN         54   среднее             1   
...         ...            ...        ...       ...           ...   
21489         2            NaN         47   Среднее             1   
21495         1            NaN         50   среднее             1   
21497         0            NaN         48    ВЫСШЕЕ             0   
21502         1            NaN         42   среднее             1   
21510         2            NaN         28   среднее             1   

               family_status  family_status_id gender  income_type  debt  \
12          гражданский брак                 1      M    пенсионер     0   
26           женат / замужем                 0      M  госслужащий     0   
29     Не женат / не замужем                 4      F    пенсионер     0   
41           женат / замужем                 0      F  госслужащий     0   
55          гражданский брак                 1      F    пенсионер     1   
...                      ...               ...    ...          ...   ...   
21489        женат / замужем                 0      M    компаньон     0   
21495       гражданский брак                 1      F    сотрудник     0   
21497        женат / замужем                 0      F    компаньон     0   
21502        женат / замужем                 0      F    сотрудник     0   
21510        женат / замужем                 0      F    сотрудник     0   

       total_income                           purpose  
12              NaN                   сыграть свадьбу  
26              NaN                       образование  
29              NaN  строительство жилой недвижимости  
41              NaN  сделка с подержанным автомобилем  
55              NaN                   сыграть свадьбу  
...             ...                               ...  
21489           NaN              сделка с автомобилем  
21495           NaN                           свадьба  
21497           NaN        строительство недвижимости  
21502           NaN  строительство жилой недвижимости  
21510           NaN           приобретение автомобиля  

[2174 rows x 12 columns]

Из таблицы видно, что пропуски в обоих столбцах допущены в одних и тех же строках. Данные пропуски можно отнести к случайным, поскольку вероятность пропуска зависит от других значений в таблице, но не от значений собственного столбца.

Чтобы ответить на поставленные  нужно избавиться от пропусков. Удалять строки нельзя, так как будет утеряны 10% данных. Будем производить замену пропусков. Для замены пропусков в столбце 'total_income' будем применять медиану и группировку значений по типу занятости.

Посмотрим, какие типы занятости указаны в таблице

In [7]:
# смотрим на уникальные значения столбца и их количество
bank['income_type'].value_counts()

сотрудник          11119
компаньон           5085
пенсионер           3856
госслужащий         1459
безработный            2
предприниматель        2
студент                1
в декрете              1
Name: income_type, dtype: int64

In [8]:
# производим группировку таблицы и подсчитаваем медиану для столбца со стажем по каждому типу занятости
bank_median = bank.groupby('income_type')['days_employed'].median()

In [9]:
# выводим таблицу со значением медиан столбца стажа для каждого типа занятости
display(bank_median)

income_type
безработный        366413.652744
в декрете           -3296.759962
госслужащий         -2689.368353
компаньон           -1547.382223
пенсионер          365213.306266
предприниматель      -520.848083
сотрудник           -1574.202821
студент              -578.751554
Name: days_employed, dtype: float64

Производим замену пропусков в столбце 'days_employed', на подсчитанные значения медиан. Для этого применим метод fillna() и transform()

In [10]:
# производим замен упропусков
bank['days_employed'] = bank['days_employed'].fillna(bank.groupby('income_type')['days_employed'].transform('median'))

Посчитам пропуски в таблице, после замены.

In [11]:
# подсчёт пропусков
bank.isna().sum()

children               0
days_employed          0
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

Пропуски в столбце 'days_employed' исчезли. Теперь произведём замену пропусков в столбце 'total_income'. Алгоритм замены, точно такой же, как и при замене пропусков в столбце 'days_employed'

In [12]:
# производим группировку таблицы и подсчитаваем медиану для столбца с доходами по каждому типу занятости
bank_median_1 = bank.groupby('income_type')['total_income'].median()

In [13]:
# выводим таблицу со значением медиан столбца доходов для каждого типа занятости
display(bank_median_1)

income_type
безработный        131339.751676
в декрете           53829.130729
госслужащий        150447.935283
компаньон          172357.950966
пенсионер          118514.486412
предприниматель    499163.144947
сотрудник          142594.396847
студент             98201.625314
Name: total_income, dtype: float64

Производим замену пропусков в столбце 'total_income', на подсчитанные значения медиан. Для этого применим метод fillna() и transform()

In [14]:
# производим замену пропусков
bank['total_income'] = bank['total_income'].fillna(bank.groupby('income_type')['total_income'].transform('median'))

Посчитам пропуски в таблице, после замены.

In [15]:
# подсчёт пропусков
bank.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

В таблице не осталось пропусков. Но нужно продолжать её обрабатывать, чтобы можно было ответить на поставленые вопросы.

Посмотрим на значения в столбце 'dob_years'

In [16]:
# смотрим на уникальные значения и их количество
bank['dob_years'].value_counts()

35    617
40    609
41    607
34    603
38    598
42    597
33    581
39    573
31    560
36    555
44    547
29    545
30    540
48    538
37    537
50    514
43    513
32    510
49    508
28    503
45    497
27    493
56    487
52    484
47    480
54    479
46    475
58    461
57    460
53    459
51    448
59    444
55    443
26    408
60    377
25    357
61    355
62    352
63    269
64    265
24    264
23    254
65    194
66    183
22    183
67    167
21    111
0     101
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

У 101 клиента возраст указан 0 лет. Предположим, что при заполнении анкеты клиенты забыли указать свой возраст, а при переносе данных в таблицу вместо пропуска указали число 0. Их доля вобщем числе строк не превышает 0,5%. Не вижу смысла заниматься их обработкой, удалим данные строки.

In [17]:
# удаляем строки, где возраст указан 0 лет
bank = bank.loc[bank['dob_years'] > 0]

In [18]:
# смотрим информацию о таблице
bank.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21424 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21424 non-null  int64  
 1   days_employed     21424 non-null  float64
 2   dob_years         21424 non-null  int64  
 3   education         21424 non-null  object 
 4   education_id      21424 non-null  int64  
 5   family_status     21424 non-null  object 
 6   family_status_id  21424 non-null  int64  
 7   gender            21424 non-null  object 
 8   income_type       21424 non-null  object 
 9   debt              21424 non-null  int64  
 10  total_income      21424 non-null  float64
 11  purpose           21424 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.1+ MB


Строки удалены. Теперь разберёмся с отрицательными значениями в столбце 'days_employed'

Изучим количество аномальных значений стажа работы 'days_employed', исходя из сведений, что единицей измерения таких данных является день. Возьмём размёр трудового стажа 45 лет, его продолжительность в днях будет составлять: 16425 дня. Соответсвенно, значения, его превышающие будут аномально высокими. Посчитаем количество таких аномальных значений.

In [19]:
display(bank[bank['days_employed'] > abs(16425)].groupby('income_type')['education'].count())

income_type
безработный       2
пенсионер      3836
Name: education, dtype: int64

Датасет содержит 3838 записи с аномальными значениями 'days_employed', из которых 3836 приходится на пенсионеров. Скорее всего ишибка в выгрузке данных. Напишем функцию days_employed_converse, которая пересчитает аномальные значения в дни, исходя из предположения, что аномалии выгружены в часах. А так же вернёт положительные значения стажа.

In [20]:
# функция проверяет значения стажа и возвращает модуль от деленяи на 24, если значение больше 16425
# или возвращает модуль числа во всех иных случаях
def days_employed_converse(row):
    if abs(row) > 16425:
        return abs(row/24)
    elif abs(row) >0 and abs(row) <= 16425:
        return abs(row)
    else:
        return row
    
# производим замену аномальных значений в столбце 'days_employed', применяя метод apply()
bank['days_employed'] = bank['days_employed'].apply(days_employed_converse)  
# взглянем на первые 10 строк таблицы
bank.head(10)

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   Среднее             1   
3         3    4124.747207         32   среднее             1   
4         0   14177.753002         53   среднее             1   
5         0     926.185831         27    высшее             0   
6         0    2879.202052         43    высшее             0   
7         0     152.779569         50   СРЕДНЕЕ             1   
8         2    6929.865299         35    ВЫСШЕЕ             0   
9         0    2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

Аномальные значения в столбце 'days_employed' заменены. 

Поскольку одним из вопросов является определения зависимости возврата кредита в срок от количества детей, посмотрим какие значения находятся в столбце 'children' и посчитаем их.

In [21]:
# смотрим на уникальные значения столбца 'children'
bank['children'].value_counts()

 0     14080
 1      4802
 2      2042
 3       328
 20       75
-1        47
 4        41
 5         9
Name: children, dtype: int64

Из полученных данных видно 2 аномалии. У 75 клиентов банка - 20 детей, а у 47 клиентов - '-1' ребёнок. Предположим, что вместо 20 детей, должно быть указано 2 ребёнка, а '-1' это один ребёнок. Напишем функцию children_converse, которая преобразует количество детей из 20 в 2, а отрицательное значение преобразует в положительные.

In [22]:
# функция проверяет значение количетсва детей и возвращает результат от деления на 10, если количетсво равно 20
# и преобразует отрицательные значения в положительные
def children_converse(row):
    if row == 20:
        row = row / 10
        return row
    elif row <0:
        row = row * -1
        return row
    else:
        return row

# производим замену аномальных значений в столбце 'children', применяя метод apply()    
bank['children'] = bank['children'].apply(children_converse)   

Ещё раз посмотрим на уникальные значения столбца 'children' и посчитаем их.

In [23]:
# уникальные значения столбца 'children'
bank['children'].value_counts()

0.0    14080
1.0     4849
2.0     2117
3.0      328
4.0       41
5.0        9
Name: children, dtype: int64

Аномальные значения в столбце 'children' заменены.

Продолжим исследовать таблицу. Посмотрим на столбец 'education'

In [24]:
# уникальные значения столбца 'education'
bank['education'].value_counts()

среднее                13691
высшее                  4686
СРЕДНЕЕ                  770
Среднее                  708
неоконченное высшее      666
ВЫСШЕЕ                   273
Высшее                   266
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
УЧЕНАЯ СТЕПЕНЬ             1
Ученая степень             1
Name: education, dtype: int64

Всего 5 типов, но отличаются регистром. Переведём типы образования к нижнему регистру методом str.lower()

In [25]:
# методом str.lower() приводим столбец 'education' к нижнему регистру
bank['education'] = bank['education'].str.lower()

In [26]:
# уникальные значения столбца 'education'
bank['education'].value_counts()

среднее                15169
высшее                  5225
неоконченное высшее      742
начальное                282
ученая степень             6
Name: education, dtype: int64

Поскольку одним из вопросов является определения зависимости возврата кредита в срок от семейного положения, посмотрим какие значения находятся в столбце 'family_status' и посчитаем их.

In [27]:
# уникальные значения столбца 'family_status'
bank['family_status'].value_counts()

женат / замужем          12331
гражданский брак          4156
Не женат / не замужем     2797
в разводе                 1185
вдовец / вдова             955
Name: family_status, dtype: int64

Аномалий в столбце с семейным положением нет. Посмотрим на половую принадлежность клиентов.

In [28]:
# уникальные значения столбца 'gender'
bank['gender'].value_counts()

F      14164
M       7259
XNA        1
Name: gender, dtype: int64

14164 клиента банка оказались женщинами, 7259 клиента оказались мужчинами. И скорее всего 1 клиент не указал пол, при заполнении анкеты. Удалим данную строку из таблицы. 

In [29]:
# удаляем строку, гдене указан пол клиента
bank = bank.loc[bank['gender'] != 'XNA']

In [30]:
# выводим информацию о таблице
bank.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21423 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21423 non-null  float64
 1   days_employed     21423 non-null  float64
 2   dob_years         21423 non-null  int64  
 3   education         21423 non-null  object 
 4   education_id      21423 non-null  int64  
 5   family_status     21423 non-null  object 
 6   family_status_id  21423 non-null  int64  
 7   gender            21423 non-null  object 
 8   income_type       21423 non-null  object 
 9   debt              21423 non-null  int64  
 10  total_income      21423 non-null  float64
 11  purpose           21423 non-null  object 
dtypes: float64(3), int64(4), object(5)
memory usage: 2.1+ MB


Наконец изучим данные в столбце 'purpose'

In [31]:
# уникальные значения столбца 'purpose'
bank['purpose'].value_counts()

свадьба                                   792
на проведение свадьбы                     773
сыграть свадьбу                           769
операции с недвижимостью                  673
покупка коммерческой недвижимости         661
покупка жилья для сдачи                   651
операции с коммерческой недвижимостью     649
операции с жильем                         647
покупка жилья                             641
жилье                                     641
покупка жилья для семьи                   640
строительство собственной недвижимости    633
операции со своей недвижимостью           630
недвижимость                              630
строительство жилой недвижимости          623
покупка недвижимости                      620
строительство недвижимости                620
покупка своего жилья                      619
ремонт жилью                              610
покупка жилой недвижимости                604
на покупку своего автомобиля              502
заняться высшим образованием      

Аномальных данных в столбце не наблюдается. Все цели написаны с пременением нижнего регистра, без спец.символов. Группировку столбца по ключевым целям будем проводить позже. Пока можно сделать выводы по проделанной работе.

**Вывод**
* были подсчитаны пропуски в столбцах 'days_employed' и 'total_income'
* пропуски были заменены медианными значениям по группировке по столбцу 'income_type'
* в столбце 'days_employed' были устранены аномалии с помощью функции 'days_employed_converse'
* аномалии в столбцах 'dob_years' 'gender' были удалены, так как их общий объем не превышал 0,5% от всего объема данных
* была написана функция 'children_converse', которая устраняла аномалии в столбце 'children'
* значения в столбце 'education' были приведены к нижнему регистру
* пропуски в данных скорее всего возникли из-за некорректной выгрузки данных, их можно отнести к случайным.

### Замена типа данных

Из данных о таблице видно, что значения в столбцах 'days_employed' и 'total_income' являются вещественными числами (float64). Переведём их значения в целые числа методом astype(), указав аргумент ('int').

In [32]:
# переведем вещественные числа в целые
bank['days_employed'] = bank['days_employed'].astype('int')
bank['total_income'] = bank['total_income'].astype('int')

In [33]:
# снова посмотрим на информацию о таблице
bank.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21423 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21423 non-null  float64
 1   days_employed     21423 non-null  int64  
 2   dob_years         21423 non-null  int64  
 3   education         21423 non-null  object 
 4   education_id      21423 non-null  int64  
 5   family_status     21423 non-null  object 
 6   family_status_id  21423 non-null  int64  
 7   gender            21423 non-null  object 
 8   income_type       21423 non-null  object 
 9   debt              21423 non-null  int64  
 10  total_income      21423 non-null  int64  
 11  purpose           21423 non-null  object 
dtypes: float64(1), int64(6), object(5)
memory usage: 2.1+ MB


Значения в столбце 'children' оказались вещественными числами. Переведём их в целые.


In [34]:
# переведем вещественные числа в целые и снова посмотрим на информацию о таблице
bank['children'] = bank['children'].astype('int')
bank.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21423 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21423 non-null  int64 
 1   days_employed     21423 non-null  int64 
 2   dob_years         21423 non-null  int64 
 3   education         21423 non-null  object
 4   education_id      21423 non-null  int64 
 5   family_status     21423 non-null  object
 6   family_status_id  21423 non-null  int64 
 7   gender            21423 non-null  object
 8   income_type       21423 non-null  object
 9   debt              21423 non-null  int64 
 10  total_income      21423 non-null  int64 
 11  purpose           21423 non-null  object
dtypes: int64(7), object(5)
memory usage: 2.1+ MB


Снова посмотрим на первые 10 строк таблицы

In [35]:
# выведем первые 10 строк таблицы
bank.head(10)

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0          14177         53   среднее             1   
5         0            926         27    высшее             0   
6         0           2879         43    высшее             0   
7         0            152         50   среднее             1   
8         2           6929         35    высшее             0   
9         0           2188         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

**Вывод**

* в трёх столбах произвели замену вещественных чисел на целые, с помощью метода astype(), указав аргумент ('int')

### Обработка дубликатов

Проверим таблицу на наличие дубликатов. Применим метод duplicated(). В сочетании с методом sum() он возвращает количество дубликатов. К нижнему регистру значения в столбце 'education' были прведены ранее.

In [36]:
bank.duplicated().sum()

71

В таблице 71 дубликат. Посмотрим на них. Применяя метод duplicated() мы получили Series с логическими значениями — теперь запросим из исходной таблицы строки, которым соответствует True в этом Series. Выведем на экран первые несколько строк с результатами:

In [37]:
# создадим таблицу из дубликатов
duplicated_bank = bank[bank.duplicated()]

In [38]:
# выведем значения на экран
display(duplicated_bank)

children  days_employed  dob_years education  education_id  \
2849          0           1574         41   среднее             1   
3290          0          15217         58   среднее             1   
4182          1           1574         34    высшее             0   
4851          0          15217         60   среднее             1   
5557          0          15217         58   среднее             1   
...         ...            ...        ...       ...           ...   
20702         0          15217         64   среднее             1   
21032         0          15217         60   среднее             1   
21132         0           1574         47   среднее             1   
21281         1           1574         30    высшее             0   
21415         0          15217         54   среднее             1   

          family_status  family_status_id gender income_type  debt  \
2849    женат / замужем                 0      F   сотрудник     0   
3290   гражданский брак                 1      F   пенсионер     0   
4182   гражданский брак                 1      F   сотрудник     0   
4851   гражданский брак                 1      F   пенсионер     0   
5557   гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
20702   женат / замужем                 0      F   пенсионер     0   
21032   женат / замужем                 0      F   пенсионер     0   
21132   женат / замужем                 0      F   сотрудник     0   
21281   женат / замужем                 0      F   сотрудник     0   
21415   женат / замужем                 0      F   пенсионер     0   

       total_income                            purpose  
2849         142594            покупка жилья для семьи  
3290         118514                    сыграть свадьбу  
4182         142594                            свадьба  
4851         118514                            свадьба  
5557         118514                    сыграть свадьбу  
...             ...                                ...  
20702        118514         дополнительное образование  
21032        118514              заняться образованием  
21132        142594                       ремонт жилью  
21281        142594  покупка коммерческой недвижимости  
21415        118514                  операции с жильем  

[71 rows x 12 columns]

Возможной причиной появления дубликатов является человеческий фатор. Предположу, что при формировании таблицы, данные некоторых клиентов могли случайно по нескольку раз добавить.
Удалим дубликаты при помощи метода drop_duplicates(). Чтобы обновить индексацию после удаления дубликатов, применим метод reset_index(), дадим ему аргумент drop=True, чтобы не создавать новую колонку индексов.

In [39]:
# удалим дубликаты и обновим индексацию данных
bank = bank.drop_duplicates().reset_index(drop=True) 

In [40]:
# посмотрим информацию о таблице
bank.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21352 entries, 0 to 21351
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21352 non-null  int64 
 1   days_employed     21352 non-null  int64 
 2   dob_years         21352 non-null  int64 
 3   education         21352 non-null  object
 4   education_id      21352 non-null  int64 
 5   family_status     21352 non-null  object
 6   family_status_id  21352 non-null  int64 
 7   gender            21352 non-null  object
 8   income_type       21352 non-null  object
 9   debt              21352 non-null  int64 
 10  total_income      21352 non-null  int64 
 11  purpose           21352 non-null  object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB


**Вывод**

* применив метод duplicated() в сочетании с методом sum(), обнаружили и сосчитали дубликаты в таблице
* удалили дубликаты методом drop_duplicates() и обновили индексацию методом reset_index


###  Лемматизация

Чтобы ответить на вопрос, как разные цели кредита влияют на его возврат в срок, применим процесс приведения слов в столбце 'purpose' к их словарной форме (лемматизация). Для этого импортируем библиотеку pymystem3. Применим метод join(), и посчитаем частоту упоминаний слов в стобце при помощи счётчика Counter, импортированного из модуля collections.

In [41]:
# 'склеим' слова в столбце 'purpose' методом join() и посчитаем леммы
m = Mystem()
text = ' '.join(bank['purpose'])
lemmas = m.lemmatize(text)
display(Counter(lemmas))         

Counter({'покупка': 5869,
         ' ': 54785,
         'жилье': 4436,
         'приобретение': 459,
         'автомобиль': 4284,
         'дополнительный': 902,
         'образование': 3995,
         'сыграть': 760,
         'свадьба': 2310,
         'операция': 2593,
         'с': 2904,
         'на': 2210,
         'проведение': 764,
         'для': 1286,
         'семья': 637,
         'недвижимость': 6327,
         'коммерческий': 1306,
         'жилой': 1224,
         'строительство': 1873,
         'собственный': 633,
         'подержать': 849,
         'свой': 2223,
         'со': 627,
         'заниматься': 900,
         'сделка': 938,
         'получение': 1309,
         'высокий': 1366,
         'подержанный': 113,
         'профильный': 435,
         'сдача': 649,
         'ремонт': 605,
         '\n': 1})

Из полученных результатов выделим 4 основные цели для получения кредита: жильё или недвижимость, автомобиль, свадьба, образование. Напишем функцию purpose_replacemente, которая будет возвращать соотвествующие значения из столбца 'purpose' в новый столбец 'purpose_category'

In [42]:
# пишем функцию purpose_replacemente, которая будет возвращать соотвествующие значения из столбца 'purpose' 
# в новый столбец 'purpose_category'
# добавим новый столбец в таблицу 
# смотрим на уникальные значения нового столбца и сосчитаем их
def purpose_replacemente(purpose):
    if 'недвиж' in purpose or 'жиль' in purpose:
        return 'недвижимость'
    elif 'авто' in purpose:
        return 'автомобиль'
    elif 'свадьб' in purpose:
        return 'свадьба'
    else:
        return 'образование'
bank['purpose_category'] = bank['purpose'].apply(purpose_replacemente)
bank['purpose_category'].value_counts()    


недвижимость    10763
автомобиль       4284
образование      3995
свадьба          2310
Name: purpose_category, dtype: int64

Посмотрим на первые 10 строк обновленной таблицы

In [43]:
# выводим первые 10 строк таблицы
bank.head(10)

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0          14177         53   среднее             1   
5         0            926         27    высшее             0   
6         0           2879         43    высшее             0   
7         0            152         50   среднее             1   
8         2           6929         35    высшее             0   
9         0           2188         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                      purpose purpose_category  
0               покупка жилья     недвижимость  
1     приобретение автомобиля       автомобиль  
2               покупка жилья     недвижимость  
3  дополнительное образование      образование  
4             сыграть свадьбу          свадьба  
5               покупка жилья     недвижимость  
6           операции с жильем     недвижимость  
7                 образование      образование  
8       на проведение свадьбы          свадьба  
9     покупка жилья для семьи     недвижимость

**Вывод**

* произвели лемматизацию значений в столбце 'purpose' и посчитали частоту упоминаний слов
* выделели 4 основные цели получения кредита
* написали функцию, которая проходится по столбцу 'purpose' и возвращает одну из 4 основных целей в новый столбец

### Категоризация данных

Для ответа на 3 оставшихся вопроса применим категоризацию данных - объединению избранных данных в произвольные группы по заданному критерию. В нашем случае данными кретериями будут: наличие детей, семейное положение и уровень дохода.

Ранее выводили уникальные значения для столбца 'children'. Сгруппируем данные, чтобы численности каждой группы хватало для формулировки выводов. Нужна категоризация — объединение данных в категории. Распределим клиентов так: отдельно бездетные, отдельно 1-2 ребёнка в семье и остальных обозначим как многодетные. К многодетным семьям отнесёс семью с 3,4,5 детьми, так как по отдельности, каждая из этих категорий имеет малый вес в общем объеме данных. Запишем правила классификации клиентов как функции. На вход функции попадает количество детей в семье, а возвращает она категорию клиента.  

In [44]:
# Пишем функцию children_category.  На вход функции попадает количество детей в семье, 
# а возвращает она категорию клиента,  в соответсвии с условием:
# - 'бездетные', если children =0;
# - '1-2 ребёнка', если children от 1 до2;
# - 'многодетные', если children >2.
# Создаём отдельный столбец 'children_category' и в его ячейках записываем значения, возвращаемые функцией.
# Применим метод apply(). Вызовем его для столбца children, так как в нём содержатся данные, которые функция примет на вход. 
# Аргументом метода станет сама функция children_category.  
# Выведем на экран первые 10 строк таблицы
def children_category(children):
    if children == 0:
        return 'бездетные'
    if children > 0 and children <3:
        return '1-2 ребёнка'
    return 'многодетные' 

bank['children_category'] = bank['children'].apply(children_category)
bank.head(10)

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0          14177         53   среднее             1   
5         0            926         27    высшее             0   
6         0           2879         43    высшее             0   
7         0            152         50   среднее             1   
8         2           6929         35    высшее             0   
9         0           2188         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                      purpose purpose_category children_category  
0               покупка жилья     недвижимость       1-2 ребёнка  
1     приобретение автомобиля       автомобиль       1-2 ребёнка  
2               покупка жилья     недвижимость         бездетные  
3  дополнительное образование      образование       многодетные  
4             сыграть свадьбу          свадьба         бездетные  
5               покупка жилья     недвижимость         бездетные  
6           операции с жильем     недвижимость         бездетные  
7                 образование      образование         бездетные  
8       на проведение свадьбы          свадьба       1-2 ребёнка  
9     покупка жилья для семьи     недвижимость         бездетные

Теперь проведём категоризацию данных по семейному положению. Выделим 4 категории: есть семья, нет семьи, в разводе, овдовел. В категорию 'нет семьи' отнесем значения: 'гражданский брак', 'Не женат / не замужем', поскольку клиент не может предоставить документы о регистрации брака, и соответственно банк, в случае просрочек платежей, не в праве предъявлять претензии гражданскому супругу. Напишем функцию family_category. На вход функции попадает семейное положение , а возвращает она категорию клиента.

In [45]:
# Пишем функцию family_category.  На вход функции попадает семейный статус, 
# а возвращает она категорию клиента,  в соответствии с условием:
# - 'нет семьи ', если family_status = 'гражданский брак', 'Не женат / не замужем;
# - 'овдовел', если family_status = вдовец / вдова ;
# - 'в разводе', если family_status = в разводе.
# - 'есть семья' во всех остальных случаях.
# Создаём отдельный столбец 'family_category' и в его ячейках записываем значения, возвращаемые функцией.
# Применим метод apply(). Вызовем его для столбца family_status, так как в нём содержатся данные, которые функция примет на вход. 
# Аргументом метода станет сама функция family_category.  
# Выведем на экран первые 10 строк таблицы
def family_category(family_status):
    if 'Не женат' in family_status or 'не замужем' in family_status or 'брак' in family_status:
        return 'нет семьи'
    elif 'вдов' in family_status:
        return 'овдовел'
    elif 'развод' in family_status:
        return 'в разводе'
    else:
        return 'есть семья'
bank['family_category'] = bank['family_status'].apply(family_category)
bank.head(10)

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0          14177         53   среднее             1   
5         0            926         27    высшее             0   
6         0           2879         43    высшее             0   
7         0            152         50   среднее             1   
8         2           6929         35    высшее             0   
9         0           2188         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                      purpose purpose_category children_category  \
0               покупка жилья     недвижимость       1-2 ребёнка   
1     приобретение автомобиля       автомобиль       1-2 ребёнка   
2               покупка жилья     недвижимость         бездетные   
3  дополнительное образование      образование       многодетные   
4             сыграть свадьбу          свадьба         бездетные   
5               покупка жилья     недвижимость         бездетные   
6           операции с жильем     недвижимость         бездетные   
7                 образование      образование         бездетные   
8       на проведение свадьбы          свадьба       1-2 ребёнка   
9     покупка жилья для семьи     недвижимость         бездетные   

  family_category  
0      есть семья  
1      есть семья  
2      есть семья  
3      есть семья  
4       нет семьи  
5       нет семьи  
6      есть семья  
7      есть семья  
8       нет семьи  
9      есть семья

Теперь проведём категоризацию данных по уровню дохода. Посмотрим на максимальное и минимальное значения в столбце.

In [46]:
# проверяем max и min  в столбце 'total_income'
bank['total_income'].max()

2265604

In [47]:
bank['total_income'].min()

20667

Разница в 100 раз! Выделим 4 категории по уровню месячного дохода: 'низкий', 'средний', 'высокий', 'безграничный'. Определим интервалы по уровню дохода для каждой из групп. Применим метод pd.qcut(), задав ему параметр q=4, для создания 4-х равных интервалов. 

In [48]:
#определяем границы интервалов
pd.qcut(bank['total_income'], q=4)

0         (195783.5, 2265604.0]
1         (107648.25, 142594.0]
2          (142594.0, 195783.5]
3         (195783.5, 2265604.0]
4          (142594.0, 195783.5]
                  ...          
21347     (195783.5, 2265604.0]
21348      (142594.0, 195783.5]
21349    (20666.999, 107648.25]
21350     (195783.5, 2265604.0]
21351    (20666.999, 107648.25]
Name: total_income, Length: 21352, dtype: category
Categories (4, interval[float64]): [(20666.999, 107648.25] < (107648.25, 142594.0] < (142594.0, 195783.5] < (195783.5, 2265604.0]]

In [49]:
# Пишем функцию total_income_category.  На вход функции попадает ежемесячный доход, 
# а возвращает она категорию клиента,  в соответсвии с условием:
# - 'низкий', если total_income < 107000;
# - 'средний', если total_income >= 107000 and total_income < 143000;
# - 'высокий', total_income >= 143000 and total_income < 196000.
# - 'безграничный' во всех остальных случаях.
# Создаём отдельный столбец 'family_category' и в его ячейках записываем значения, возвращаемые функцией.
# Применим метод apply(). Вызовем его для столбца family_status, так как в нём содержатся данные, которые функция примет на вход. 
# Аргументом метода станет сама функция family_category.  
# Выведем на экран первые 10 строк таблицы
def total_income_category(total_income):
    if total_income < 107000:
        return 'низкий'
    if total_income >= 107000 and total_income < 143000:
        return 'средний'
    if total_income >= 143000 and total_income < 196000:
        return 'высокий'
    return 'безграничный'
bank['total_income_category'] = bank['total_income'].apply(total_income_category)
bank.head(10)

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0          14177         53   среднее             1   
5         0            926         27    высшее             0   
6         0           2879         43    высшее             0   
7         0            152         50   среднее             1   
8         2           6929         35    высшее             0   
9         0           2188         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                      purpose purpose_category children_category  \
0               покупка жилья     недвижимость       1-2 ребёнка   
1     приобретение автомобиля       автомобиль       1-2 ребёнка   
2               покупка жилья     недвижимость         бездетные   
3  дополнительное образование      образование       многодетные   
4             сыграть свадьбу          свадьба         бездетные   
5               покупка жилья     недвижимость         бездетные   
6           операции с жильем     недвижимость         бездетные   
7                 образование      образование         бездетные   
8       на проведение свадьбы          свадьба       1-2 ребёнка   
9     покупка жилья для семьи     недвижимость         бездетные   

  family_category total_income_category  
0      есть семья          безграничный  
1      есть семья               средний  
2      есть семья               высокий  
3      есть семья          безграничный  
4       нет семьи               высокий  
5       нет семьи          безграничный  
6      есть семья          безграничный  
7      есть семья               средний  
8       нет семьи                низкий  
9      есть семья               высокий

In [50]:
bank['total_income_category'].value_counts() 

средний         5556
безграничный    5325
низкий          5279
высокий         5192
Name: total_income_category, dtype: int64

**Вывод**

* провели категоризацию данных по 3 критериям отдельно (семейное положение, наличие детей, ежемесячный доход)
* каждую категорию разбивали на 4 подкатегории
* добавили в таблицу 3 новых столбца с результатами категоризации.

Можно приступать к группировке таблицы, чтобы ответить на поставленные вопросы.

## Ответьте на поставленные вопросы

Для ответа на поставленые вопросы нужно создать сводные таблицы. Для этого применяем метод pivot_table(). В зависимости от поставленного вопроса будут меняться входные параметры index(столбцы, по которым группируют данные). В качестве аргумента values(значения, по которым мы хотим увидеть сводную таблицу) везде будт значения столбца 'debt'. Аргумент функции aggfun(функция, применяемая к значениям), во всех случаях будет 'mean'. Проведем исследования. 

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [51]:
data_pivot_children = bank.pivot_table(index=['children_category'], values='debt', aggfunc='mean')
data_pivot_children

debt
children_category          
1-2 ребёнка        0.092622
бездетные          0.075458
многодетные        0.082011

**Вывод**

Наличие детей сказывается на возврат кредита в срок. Так у бездетных ниже процент просроченных платежей по кредиту, а у клиентов с 1-2 детьми процент просроченных платежей даже выше, чем у многодетных. Как известно, с появлением ребенка затраты резко увеличиваются. Поэтому бездетные реже допускают просрочки по платежам. Многодетные уже знают на своём примере, через какие траты им придётся пройти, поэтому реже допускают просрочек относительно семей с 1-2 ребёнками.

- Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [52]:
data_pivot_family_status = bank.pivot_table(index=['family_category'], values='debt', aggfunc='mean')
data_pivot_family_status

debt
family_category          
в разводе        0.071730
есть семья       0.075427
нет семьи        0.095190
овдовел          0.064990

**Вывод**

Клиенты, которые овдовели, меньше допускают просрочек по платежам. Больше всего просрочек у клиентов, которые не состоят в браке. Клиенты у которых есть/была семья примерно одинаково допускают просрочки по платежам.

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [53]:
data_pivot_total_income = bank.pivot_table(index=['total_income_category'], values='debt', aggfunc='mean')
data_pivot_total_income

debt
total_income_category          
безграничный           0.071737
высокий                0.084553
низкий                 0.079939
средний                0.088193

**Вывод**

Больше всего просрочек у клиентов со средним уровнем дохода, но меньше всего просрочек у клиентов с безграничным доходом. Клиенты с низким достатком платят по кредиту лучше, чем со средним и высоким. Предположу, что люди с низким уровнем дохода более финансово ответственны, и прежде чем подать заявку на кредит, просчитают сколько они могут себе позволить выделить из бюджета на ежемесячное погашение кредита. Люди со средним и высоким достатком считают, что их дохода вполне хватит, чтобы спокойно гасить платежи и они могут быть не готовы к непредвиденным расходам(поломка машины, бытовой техники, срочный ремонт и т.д.)

- Как разные цели кредита влияют на его возврат в срок?

In [54]:
data_pivot_purpose = bank.pivot_table(index=['purpose_category'], values='debt', aggfunc='mean')
data_pivot_purpose

debt
purpose_category          
автомобиль        0.093371
недвижимость      0.072378
образование       0.092616
свадьба           0.079654

**Вывод**

Кредиты, которые берутся с целью проведения свадьбы и операций с недвижимостью имеют меньший процент просрочек. При покупке автомобиля или оплаты образования процент просрочек примерно одинаков.

## Общий вывод

Проведя исследование были даны ответы на поставленые вопросы:
1. Есть зависимость между наличием детей и возвратом кредита в срок.
2. Есть зависимость между семейным положением и возвратом кредита в срок.
3. Есть зависимость между уровнем дохода и возвратом кредита в срок.
4. Цели кредита влияют на его возврат в срок.

Однако нужно заметить, что зависимость в каждой категории едва заметна. Осмелюсь утверждать, что если бы не проверялась зависимость между уровнем дохода и возвратом кредита в срок, а категоризированный столбец по уровню доходов добавлялся бы ещё одним аргументов в графу индекс при формировании сводных таблиц, то при получении 'широкой' таблицы зависимости были бы более явными.